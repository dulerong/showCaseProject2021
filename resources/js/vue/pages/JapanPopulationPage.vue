<template lang="pug">
div
  h1 {{ $t("japanPopulation.title") }}
  v-row(no-gutters)
    v-select(label="prefectures" :items="$_prefectures" v-model="prefectureSelected" item-text="name" item-value="code" return-object multiple chips clearable)
  Chart(:chartData="chartData")
</template>

<script>
import axios from 'axios'
import { mapMutations } from 'vuex'
export default {
  components: {
    Chart: require('~components/Chart').default
  },
  data: () => ({
    prefectureSelected: [],
    error: false,
    apiResponse: null,
    chartData: []
  }),
  watch: {
    prefectureSelected: async function (val, oldVal) {
      if (val.length < oldVal.length) {
        const deletedPrefecture = oldVal.find(item => !val.includes(item))
        this.removePrefecture(deletedPrefecture)
      } else {
        const newPrefecture = val.find(item => !oldVal.includes(item))
        await this.fetchData(newPrefecture)
        
        if (this.error) return
        this.chartData.push({ name: newPrefecture.name, data: this.apiResponse})
      }
    },
  },
  methods: {
    ...mapMutations('notification', ['showNotification']),
    fetchData: async function(prefecture) {
      const API_ADDRESS = "https://opendata.resas-portal.go.jp/api/v1/population/composition/perYear"
      const params = { prefCode: prefecture.code }
      const API_KEY = process.env.NODE_ENV === 'test' ? 'API_KEY' : process.env.MIX_RESAS_KEY
      const headers = {
        "Content-Type": "application/json",
        "X-API-KEY": API_KEY
      }
      return axios.get(API_ADDRESS, { params, headers })
        .then(res => this.apiResponse = res.data.result.data[0].data.map(item => item.value))
        .catch(() => {
          this.error = true
          this.showNotification({ message: 'Error occurred fetching data', color: 'error' })
        })
    },
    removePrefecture(deletedPrefecture) {
      this.chartData = this.chartData.filter(item => item.name !== deletedPrefecture.name)
    }
  }
}
</script>