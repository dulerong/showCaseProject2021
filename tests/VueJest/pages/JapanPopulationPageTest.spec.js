import { mount, createLocalVue } from '@vue/test-utils'
import Vuetify from 'vuetify'
import Vuex from 'vuex'

import JapanPopulationPage from '~pages/JapanPopulationPage'
import prefectures from '~js/plugins/japanPrefectures'

let url = ''
let config = {}
let mockError = false

const mockResult = {
  data: {
    result: {
      data: [
        {
          data: [
            { value: 1 },
            { value: 2 },
            { value: 3 },
          ]
        }
      ]
    }
  }
}

jest.mock('axios', () => ({
  get: (_url, _config) => {
    return new Promise((resolve, reject) => {
      if (mockError) reject()

      url = _url
      config = _config
      resolve(mockResult)
    })
  }
}))

describe('Japan Population Page', () => {
  const localVue = createLocalVue()
  localVue.use(Vuex)
  localVue.$_prefectures = prefectures

  const mutations = { showNotification: jest.fn() }
  const mockStore = new Vuex.Store({
    modules: {
      notification: {
        namespaced: true,
        mutations,
      },
    },
  })

  let vuetify

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  const mountFunction = options => {
    return mount(JapanPopulationPage, {
      localVue,
      vuetify,
      mocks: {
        $t: (msg) => msg,
        $store: mockStore
      },
      stubs: {
        ApexChart: true,
      },
      ...options
    })
  }

  it('Page renders properly', () => {
    const wrapper = mountFunction()

    expect(wrapper.exists()).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Watch prefectureSelected: upon change will trigger fetchData method with correct argument', async () => {
    const fetchDataMethod = jest.spyOn(JapanPopulationPage.methods, 'fetchData')
    const prefectureSelected = { name: '北海道', code: 1 }
    const wrapper = mountFunction()
    expect(fetchDataMethod).not.toHaveBeenCalled()

    await wrapper.setData({ prefectureSelected: [prefectureSelected] })
    expect(fetchDataMethod).toHaveBeenCalledWith(prefectureSelected)
  })

  it('Watch prefectureSelected: triggers removePrefecture method when user removes selected prefectures', async () => {
    const removePrefectureMethod = jest.spyOn(JapanPopulationPage.methods, 'removePrefecture')
    const prefectureSelected = [
      { name: 'California' },
      { name: 'Washington' },
    ]
    const data = () => ({ prefectureSelected })
    const wrapper = mountFunction({ data })

    await wrapper.setData({ prefectureSelected: prefectureSelected.slice(-1) })

    expect(removePrefectureMethod).toHaveBeenCalledWith(prefectureSelected[0])
  })

  it('Method removePrefecture: removes prefecture from chartData', async () => {
    const chartData = [
      { name: 'California' },
      { name: 'Washington' },
    ]
    const data = () => ({ chartData })
    const wrapper = mountFunction({ data })

    const deletedPrefecture = chartData[0]
    await wrapper.vm.removePrefecture(deletedPrefecture)
    
    const remainingPrefecture = chartData.filter(item => item.name !== deletedPrefecture.name)
    expect(wrapper.vm.chartData).toEqual(remainingPrefecture)
  })

  it('Method fetchData: triggers axios GET request with correct parameters and return correct result', async () => {
    const wrapper = mountFunction()
    const prefectureSelected = { name: '北海道', code: 1 }

    const API_ADDRESS = "https://opendata.resas-portal.go.jp/api/v1/population/composition/perYear"
    const params = { prefCode: prefectureSelected.code }
    const headers = {
      "Content-Type": "application/json",
      "X-API-KEY": 'API_KEY'
    }

    await wrapper.vm.fetchData(prefectureSelected)
    expect(url).toBe(API_ADDRESS)
    expect(config).toStrictEqual({params, headers})
    expect(wrapper.vm.apiResponse).toEqual([1, 2, 3])
  })

  it('Method fetchData: upon axios GET failure, triggers showNotification', async () => {
    const wrapper = mountFunction()
    const prefectureSelected = { name: '北海道', code: 1 }
    
    mockError = true

    await wrapper.vm.fetchData(prefectureSelected)
    expect(mutations.showNotification).toHaveBeenCalled()
  })
})