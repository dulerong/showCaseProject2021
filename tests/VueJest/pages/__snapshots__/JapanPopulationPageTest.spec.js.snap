// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Japan Population Page Page renders properly 1`] = `
<div>
  <h1>Japan Population Page</h1>
  <div class="row no-gutters">
    <div class="v-input theme--light v-text-field v-select v-select--chips v-select--is-multi">
      <div class="v-input__control">
        <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-4" class="v-input__slot">
          <div class="v-select__slot"><label for="input-4" class="v-label theme--light" style="left: 0px; position: absolute;">prefectures</label>
            <div class="v-select__selections"><input id="input-4" readonly="readonly" type="text" aria-readonly="false" autocomplete="off"></div>
            <div class="v-input__append-inner">
              <div></div>
            </div>
            <div class="v-input__append-inner">
              <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
            </div><input type="hidden">
          </div>
          <div class="v-menu">
            <!---->
          </div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
